tracing-core/src/callsite.rs

```diff
@@ -169,3 +187,159 @@ impl Hash for Identifier {
         (self.0 as *const dyn Callsite).hash(state)
     }
 }
+
+// ===== impl Registration =====
+
+impl<T> Registration<T> {
+    /// Construct a new `Registration` from some `&'static dyn Callsite`
+    pub const fn new(callsite: T) -> Self {
+        Self {
+            callsite,
+            next: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+}
+
+impl fmt::Debug for Registration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Registration").finish()
+    }
+}
+
+// ===== impl LinkedList =====
+
+/// An intrusive atomic push only linked-list.
+struct LinkedList {
+    head: AtomicPtr<Registration>,
+}
+
+impl LinkedList {
+    fn new() -> Self {
+        LinkedList {
+            head: AtomicPtr::new(ptr::null_mut()),
+        }
+    }
+
+    fn for_each(&self, mut f: impl FnMut(&'static dyn Callsite)) {
+        let mut head = self.head.load(Ordering::Acquire);
+
+        while !head.is_null() {
+            let reg = unsafe { &*head };
+            f(reg.callsite);
+
+            head = reg.next.load(Ordering::Acquire);
+        }
+    }
+
+    fn push(&self, cs: &'static dyn Callsite) {
+        let mut head = self.head.load(Ordering::Acquire);
+
+        loop {
+            let registration = cs.registration() as *const _ as *mut _;
+
+            cs.registration().next.store(head, Ordering::Release);
+
+            match self.head.compare_exchange(
+                head,
+                registration,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            ) {
+                Ok(old) => {
+                    assert_ne!(
+                        cs.registration() as *const _,
+                        old,
+                        "Attempting to push a `Callsite` that already exists. \
+                        This will cause an infinite loop when attempting to read from the \
+                        callsite cache. This is likely a bug! You should only need to push a \
+                        `Callsite` once."
+                    );
```

Comment:
it occurs to me that it is probably more correct to make this assertion *before* the `compare_exchange`, rather than after. if we make the assertion after the compare_exchange succeeds, we will have already created the infinite loop. if the panic is caught, and the program continues executing, we'll have an infinite loop in the registry. 

can we move this to the top of the loop, before we store the head value in our next pointer (around line 239)? that way, we'll panic *before* a potential infinite loop is created, rather than making one and then panicking.